components/camera/rtsp/rtsp.go

```diff
@@ -0,0 +1,281 @@
+
+func init() {
+	registry.RegisterComponent(camera.Subtype, model, registry.Component{
+		Constructor: func(ctx context.Context, _ registry.Dependencies, cfg config.Component, logger golog.Logger) (interface{}, error) {
+			attrs, ok := cfg.ConvertedAttributes.(*Attrs)
+			if !ok {
+				return nil, utils.NewUnexpectedTypeError(attrs, cfg.ConvertedAttributes)
+			}
+			return NewRTSPCamera(ctx, attrs, logger)
+		},
+	})
+
+	config.RegisterComponentAttributeMapConverter(
+		camera.Subtype,
+		model,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf Attrs
+			attrs, err := config.TransformAttributeMapToStruct(&conf, attributes)
+			if err != nil {
+				return nil, err
+			}
+			result, ok := attrs.(*Attrs)
+			if !ok {
+				return nil, utils.NewUnexpectedTypeError(result, attrs)
+			}
+			return result, nil
+		},
+		&Attrs{},
+	)
+}
+
+// Attrs are the config attributes for an RTSP camera model.
+type Attrs struct {
+	Address          string                             `json:"rtsp_address"`
+	IntrinsicParams  *transform.PinholeCameraIntrinsics `json:"intrinsic_parameters,omitempty"`
+	DistortionParams *transform.BrownConrady            `json:"distortion_parameters,omitempty"`
+}
+
+// Validate checks to see if the attributes of the model are valid.
+func (at *Attrs) Validate() error {
+	if prefix := strings.HasPrefix(at.Address, "rtsp://"); !prefix {
+		return errors.New(`rtsp_address must begin with "rtsp://"`)
+	}
+	if err := at.IntrinsicParams.CheckValid(); err != nil {
+		return err
+	}
+	if err := at.DistortionParams.CheckValid(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// rtspCamera contains the rtsp client, and the reader function that fulfills the camera interface.
+type rtspCamera struct {
+	gostream.VideoReader
+	u                       *url.URL
+	client                  *gortsplib.Client
+	cancelCtx               context.Context
+	cancelFunc              context.CancelFunc
+	activeBackgroundWorkers sync.WaitGroup
+	latestFrame             atomic.Value
+	logger                  golog.Logger
+}
+
+// Close closes the camera.
+func (rc *rtspCamera) Close(ctx context.Context) error {
+	var clientTerminated liberrors.ErrClientTerminated
+	rc.cancelFunc()
+	rc.activeBackgroundWorkers.Wait()
+	if err := rc.client.Close(); err != nil && !errors.Is(err, clientTerminated) {
+		return err
+	}
+	return nil
+}
+
+// clientReconnectBackgroundWorker checks every 5 sec to see if the client is connected to the server, and reconnects if not.
+func (rc *rtspCamera) clientReconnectBackgroundWorker() {
+	rc.activeBackgroundWorkers.Add(1)
+	ticker := time.NewTicker(5 * time.Second)
+	goutils.ManagedGo(func() {
+		for {
+			select {
+			case <-rc.cancelCtx.Done():
+				return
+			case <-ticker.C:
+				// check if the client is still connected with an OPTIONS request
+				res, err := rc.client.Options(rc.u)
+				if err != nil && (errors.Is(err, liberrors.ErrClientTerminated{}) ||
+					strings.Contains(err.Error(), "EOF") ||
```

Comment:
nit: assign Error() to a variable first instead of calling it each time. Have you also checked that these error can't be used withe errors.Is()? EOF for example is not io.EOF?

Reply:
replaced with the errors available in syscall